internal/pkg/nftTrade-scrapers/opensea.go

```diff
@@ -0,0 +1,454 @@
+
+	filter, err := s.contract.FilterOrdersMatched(&bind.FilterOpts{Start: startBlockNum, End: &endBlockNum}, nil, nil, nil)
+	if err != nil {
+		log.Errorf("unable to create ordersmatched filter: %s", err.Error())
+		return err
+	}
+
+	for filter.Next() {
+		if err := filter.Error(); err != nil {
+			log.Errorf("unable to parse filtered event: %s", err.Error())
+			return err
+		}
+
+		ev := filter.Event
+
+		currBlockNum := new(big.Int).SetUint64(ev.Raw.BlockNumber)
+
+		if s.lastBlockNum.Cmp(currBlockNum) != 0 {
+			s.lastLogIndex = nil
+		}
+
+		s.lastBlockNum = currBlockNum
+
+		// skip if the event has already been processed
+		if s.lastLogIndex != nil && ev.Raw.Index <= *s.lastLogIndex {
+			log.Tracef("event(block: %d, log: %d, tx: %s) skipped because it has already been processed", ev.Raw.BlockNumber, ev.Raw.Index, ev.Raw.TxHash.Hex())
+			continue
+		}
+
+		// skip if the event removed due to chain reorganization
+		if ev.Raw.Removed {
+			log.Tracef("event(block: %d, log: %d, tx: %s) skipped due to chain reorganization", ev.Raw.BlockNumber, ev.Raw.Index, ev.Raw.TxHash.Hex())
+			continue
+		}
+
+		transfers, err := s.findERC721Transfers(ctx, ev.Raw.TxHash)
+		if err != nil {
+			log.Errorf("unable to find transfers of the event(block: %d, log: %d, tx: %s): %s", ev.Raw.BlockNumber, ev.Raw.Index, ev.Raw.TxHash.Hex(), err.Error())
+			return err
+		}
+
+		// skip if the event has no transfer
+		if len(transfers) == 0 {
+			log.Tracef("event(block: %d, log: %d, tx: %s) skipped due to it has no erc721 transfer log", ev.Raw.BlockNumber, ev.Raw.Index, ev.Raw.TxHash.Hex())
+			s.lastLogIndex = &ev.Raw.Index
+			continue
+		}
+
+		// skip if the event has multiple transfers due to we can't calculate the price of trade
+		if len(transfers) > 1 {
+			log.Tracef("event(block: %d, log: %d, tx: %s) skipped due to it has multiple erc721 transfer logs", ev.Raw.BlockNumber, ev.Raw.Index, ev.Raw.TxHash.Hex())
+			s.lastLogIndex = &ev.Raw.Index
+			continue
+		}
+
+		if err := s.notifyTrade(ev, transfers[0]); err != nil {
+			if !errors.Is(err, errOpenSeaShutdownRequest) {
+				log.Warnf("event(block: %d, log: %d, tx: %s) couldn't processed: %s", ev.Raw.BlockNumber, ev.Raw.Index, ev.Raw.TxHash.Hex(), err.Error())
+			}
+
+			return err
+		}
+
+		// store the log index of last processed event
+		s.lastLogIndex = &ev.Raw.Index
+	}
+
+	s.lastBlockNum = new(big.Int).SetUint64(endBlockNum + 1)
+	s.lastLogIndex = nil
+
+	return nil
+}
+
+func (s *OpenSeaScraper) notifyTrade(ev *opensea.ContractOrdersMatched, transfer *erc721Transfer) error {
+	nft, err := s.tradeScraper.datastore.GetNFT(transfer.NFTAddress.Hex(), dia.ETHEREUM, transfer.TokenID.String())
+	if err != nil {
+		// TODO: if the error indicates the nft could not found in rdb,
+		//       maybe we can create a new nft record by method SetNFT.
+		//       otherwise, the scraper will stuck at this event due to
+		//       undefined nft. @jppade, what is the desired behaivor?
```

Comment:
I think it's a good idea to SetNFT in case the nft could not be found. In fact, in the worst case we would also have to create a new NFTClass record. This is beyond the scope of this task though. You can create the additional functionality if you like. In that case I would adjust the value of the underlying bounty.